### ISSUE TITLE
Test metadata annotation

### ISSUE THREAD

**Issue by tsiq-brian** [twosigma/webtau#488]
Our integration tests are authored and maintained by different teams. Tests are currently organized by functionality, and there is no simple functionality->team association.

We'd like some way of injecting static metadata into the tests so that we can use it in reporting and somehow ensure that all tests have an owner (or fail somewhere if not). The use cases we have are...

* propagating test ownership data (a string, array of strings, or something) to the report json that gets generated so that we can notify those specific teams when their integration test fails.
* enforcing each test has ownership data associated so that no one can push a test that isn't owned by someone

For the JUnit Java/Groovy route, we can create a custom annotation and annotate the test classes/methods. For the two use cases, we can probably scan the classpath for those annotations and have a custom reporter and enforcer unit test pull that metadata.

However, most of our tests use the Groovy DSL. Maybe it's my lack of Groovy knowledge, but I don't know if there's a straightforward way to "annotate" the `scenario` call in the test, or have that data available within other tests.

**Comment by tsiq-clemens** [comment]
The report json is custom-generated, so we need
* some metadata object attachable to a `WebTauTest`
* a way to retrieve that object from the `WebTauTest` so that custom generators can do something with it

**Comment by tsiq-clemens** [comment]
For the "fail somewhere if not part", the idea is that either 
* a custom webtau test fails if not all tests have an owner (but what about this custom test's owner?) 
* the webtau suite as a whole fails
* some process that runs the webtau suite fails

**Comment by MykolaGolubyev** [comment]
Interesting request! Thank you.
You can define custom values in config. Config values should be part of a report. Owner custom config value could be it.

There are also multiple listeners one can attach custom code to. You can have custom httpHeaderValues injected into each call. You can listen for each http call and react on it.

So if each team has its own webtau config file then add a new `owner` value there.
As part of http listeners you can listen to a call, and exit hard if `cfg.owner` is not defined. 
Registering http calls listener is only available right now through `META-INF/services`. Potentially you could “attach” a jar file with impl during test run on your side to enforce the rule.  Or you can provide a company wide wrapper to run `webtau` that enforces the rule.

In custom generated report you can access `owner` value using `cfg`.

**Comment by tsiq-clemens** [comment]
More background:
We have a single "api integration tests" suite with a single webtau.cfg file, maintained by multiple teams.
The multiple-config-files approach won't work in that case.

**Comment by tsiq-karold** [comment]
This is interesting, I quite like this idea.  I can envisage how this would work for annotations (in the english, not the Java sense) at the test file/Junit class level, I think it'll be a bit harder to do nicely at the scenario level.

We could have a test metadata class which resembles something like this (pseudocode):
```
class TestMetadata(
  String name
  String description
  Map<String, Object> additionalMetadata
)
```

For Groovy DSL, this could be accomplished with a block like this in a test file:
```
metadata {
  name = "This is a better description than the file name will allow"
  description = "These scenarios set up foo, walk through this workflow, assert that this works, etc."
  additionalMetadata = {
    owner = "John"
    team = "Service A"
  }
}
```

Similarly for a Junit test class:
```
@TestMetadata(name = "", description = "", ...)
@Webtau
class MyWebTauTest {
...
```

Internally we can associate that with all relevant test objects and make it available in the report.

**Comment by MykolaGolubyev** [comment]
Good approach @tsiq-karold.
We could try to reduce nest level at least in groovy.
In java we potentially could use the same `@Webtau` annotation to pass any meta data. At least in junit5.

on a scenario level in groovy we could do something like 

```
scenario('blah') {
    meta {
        owner = "John"
    }
}
```

```
scenario('blah') {
    meta = [
        owner:  "John"
    ]
}
```

```
scenario('blah', [owner:  "John"]) {
```

We could do the same on a top level

```
meta = [
     owner:  "John"
]

scenario('blah') {
}
```

Also what about renaming `meta` to `webtau`?

```
webtau = [
     owner:  "John"
]

scenario('blah') {
}
```

**Comment by tsiq-clemens** [comment]
I think "meta" conveys the concept of "metadata" better than "webtau".
In groovy, having a map as optional 2nd argument to `scenario` feels more intuitive to me; it's also harder to get confused with stuff inside the scenario closure.

**Comment by tsiq-karold** [comment]
I don't mind reusing the `@Webtau` annotation if it looks like `@Webtau(@TestMetadata(owner = "foo", ...))`.  Wouldn't want owner, etc. to be properties of `@Webtau` for the reasons Clemens outlined.

**Comment by MykolaGolubyev** [comment]
Do you mean to hardcode `owner` in `TestMetadata`?
How about 
```
@WebTau(meta = {
        @Meta(key = "owner", value = "Clemens"),
        @Meta(key = "custom", value = "entry")})
```

Alternatively we can make a "hack" and let users use their own annotations like `@Owner("teamA")` but we will teach junit runners to read any annotation on a class and test level to add their name and corresponded value to meta data. Maybe even provide a few out of the box.

```
@Test
@DisplayName("query by first name")
@Owner("team A")
void queryByFirstName() {
```

**Comment by MykolaGolubyev** [comment]
@tsiq-clemens what about

```
webtau = {
    meta = []
    otherPotentialConfig = ...
}
```

**Comment by tsiq-clemens** [comment]
I feel this is getting somewhere... Keeping junit and groovy close to "same" would be great.

How about we define meta attributes in the `webtau.cfg`? Then we have an option of validating them and the burden of defining your own is minimized.

**Comment by MykolaGolubyev** [comment]
Could you elaborate with an example?

**Comment by tsiq-clemens** [comment]
```
testMetadata {
  owner, required = true
  team
}
```
sets up `owner` as required metadata, and `team` as optional. `required=true` here means that every test has to have an owner. 
Other metadata are not recognized / allowed.

**Comment by MykolaGolubyev** [comment]
Hmm that may not work well with the approach I outlined above. Owner is just a function. 

We still can enforce what is required in a custom test listener that you can add. In the listener you will fail hard if a test doesn’t have meta you need. 

As owner is just a simple function anyone can create other similar functions. I don’t think you can stop it.

**Comment by tsiq-clemens** [comment]
Custom test listener works for me. 
Can the `owner` / `team` functions as you outlined above be declared in webtau.cfg ? That'll put them into a single place which is better for maintenance. It may not stop people from defining others, but we could still recommend a single place of definition.

**Comment by MykolaGolubyev** [comment]
not in a way I know. Since groovy compiler won't treat webtau.cfg as a class.
Just realized that using `owner` without any prefixes won't work out of the box. 
We have two choices:
1. Specify in a config file what classes to implicitly static import (we already do implicit static imports for `scenario` and matchers)
2. Do an explicit `Meta.owner` and have it be defined in `Meta.groovy` (naming and file locations is up to test developers. Webtau won't know anything about it)

**Comment by tsiq-clemens** [comment]
2 feels easier on the framework (and probably test developer).
How would you be able to use it in junit5 and groovy?

**Comment by MykolaGolubyev** [comment]
Junit will use annotations. Will treat any annotation with a single value like a meta.

**Comment by tsiq-clemens** [comment]
Can you summarize what Junit and groovy usages will look like? We can agree on those and then dive into implementation details.

**Comment by MykolaGolubyev** [comment]
**JUnit**
```
@Webtau
@Owner("team a")
class MyTest {
   @Test
   @Owner("another team")
   void concreteTest() {
   }
}
``` 

**Groovy runner**
```
static import Meta.*

owner("team a") 

scenario("one") {}

scenario("two") {
   owner("team b")
   //....
}
```
where `Meta.owner` is defined as

```
class Meta {
   static void owner(String name) {
       associateMetaValue("owner", name)
   }
}
```

**Comment by tsiq-clemens** [comment]
For completeness' sake:
* in the groovy runner, can the `owner` method be called anywhere in the block or does it have to be at the beginning? 
* what happens if you (accidentally) have 2 `owner` calls? last one wins?

**Comment by MykolaGolubyev** [comment]
`owner` can be called anywhere, last one wins. We could throw.

**Comment by tsiq-clemens** [comment]
Maybe the developer can make that call if the framework provides `associateMetaValue("owner", name)` vs `associateFinalMetaValue("owner", name)`

**Comment by MykolaGolubyev** [comment]
Makes sense. I can work on groovy runner. And later someone can implement junit ones.

**Comment by tsiq-clemens** [comment]
go for it!

**Comment by tsiq-clemens** [comment]
As discussed in the PR, we'd like to enable this closure syntax as "annotation" for a groovy test file.

```
metadata {
  team = "foo",
  owner = "blah",
...
}
```

**Comment by MykolaGolubyev** [comment]
Are there any benefits compared to say 
```
metadata(team: "foo", owner: "blah)?
```

What I liked about `owner("team a")` is its closeness to Java proposed syntax `@Owner("team a")`. 

Not a big fan of 
```
@WebTau(meta = {
        @Meta(key = "owner", value = "Clemens"),
        @Meta(key = "custom", value = "entry")})
```

In Java I actually would prefer to be able to do something like `@TeamA`, hence the groovy introduced approach that let you do `teamA()` kind of calls. 

Note that if `metadata(team: "foo", owner: "blah)` generic one is available then you can go as loosely defined meta or as strictly defined meta as you want. The closure one is syntax sugar on top of it which we also can introduce. But I wouldn't want to have two different syntaxes. That's why I ask about closure benefits. It could be that you plan to use `delegates` or it could be just the preferred way to define it. 

Once we go closure it won't be convenient to use `keys` as variables. While `team` in that closure looks like variable, it is just a string value without quotes, i.e. no autocompletion or validation will be available. 

In the related PR I asked to get back to the main use case. If it is to associate finite number of key value pairs with a large number of test I would suggest to go with something strongly typed. 
If key-values are very different for each test, then loosely defined key value pair should work.

**Comment by tsiq-karold** [comment]
I'm personally not tied to this being a closure per se but my preference is an approach which allows me to define the metadata in a single call.  So your suggestion of `metadata(team: "foo")` is also fine with me.

I agree there's definite benefit to strong typing though.  We could do something that gets us closer to that by having `metadata()` take a vararg of `interface MetadataItem` or something like that.  So everything would look like:
```
interface MetadataItem {}

class Owner implements MetadataItem {
  Owner(String name) {...}
}

// Scenario file below
metadata(Owner("you"), Team("A-team"))

scenario("do stuff") {...}
```

I *think* IntelliJ will suggest `Owner` and `Team` when you ctrl+space inside `metadata()`.

**Comment by tsiq-karold** [comment]
I can't comment too much on the use case though, I don't have one 😄 Will let @tsiq-clemens to opine on that.

**Comment by tsiq-clemens** [comment]
+1 to 
* "Single call is better than multiple calls"
* take advantage of strong types, we expect fairly few (1-5)

How would one be able to prevent (throw at) multiple invocations outside or <gasp> inside scenarios? The main use case is "declare metadata at the file level, override at the scenario level".
I'm happy to stick overrides into optional scenario arguments, e.g.
`scenario("special sauce test", Owner("B-team")) { ...} `
which keeps them in a single fixed place

**Comment by MykolaGolubyev** [comment]
I thought about it a bit more (hiking helps :) ).
If we expose `metadata(Map)` than anyone is free to implement on top of it whatever shortcuts they desire. 
Strong types advantage is auto completion and auto validation.  Same potentially can be achieved by custom methods or any other approach like builder pattern. 

e.g.
```
def supportMetadata(StrongTypes... strongMetadata) {
    def map = collectMapFrom(strongMetadata)
    metadata(map)
}
```

@tsiq-clemens do you have an idea of what those 1-5 are? May help to come up with a better DSL for your case.

@tsiq-karold you may need to do `new Owner("you")` instead of just `Owner` in Groovy (I assume you did some Kotlin syntax). You can bypass `new` by using `@Newify` everywhere you want to use `Owner(...)` instead of `new Owner(...)`. Either way it is extra noise. Not sure how critical it is.

I think we should not introduce a second parameter to `scenario`. The reason is we may introduce some additional features in the future. If we stick `metadata` to `scenario` chances are we will want to stick that new feature there as well, and then order of args will be confusing. 

What's left is to detect if a metadata is being overriden. I am not sure how much of a problem it actually is. There are potential scenarios where one may even set metadata based on if-else logic. I.e. override within a scenario. Do we expect a problem with someone calling `metadata` multiple times?

In the PR update I made `metadata` return a previously associated values. In your own wrapper you could potentially fail hard if the data was already set. Unfortunately it will be triggered for a regular `scenario` override case. 
I am pretty sure we can figure this out and fail on a core `metadata(Map)` level. Just not sure we have to.

**Comment by tsiq-karold** [comment]
> @tsiq-karold you may need to do new Owner("you") instead of just Owner in Groovy (I assume you did some Kotlin syntax). You can bypass new by using @Newify everywhere you want to use Owner(...) instead of new Owner(...). Either way it is extra noise. Not sure how critical it is.

It was pseudocode 😁 

> What's left is to detect if a metadata is being overriden. I am not sure how much of a problem it actually is.

My main concern with this is the scenario I outlined in my last post (pseudocode again...):
```
owner("foo")
scenario("do stuff owned by foo") {...}

owner("bar")
scenario("do stuff owned by bar") {...}
```

I *think* this would result in both scenarios having owner bar which is not what one might expect.

**Comment by MykolaGolubyev** [comment]
Fair :) so that’s a good example. I agree with it. I think in this case it is relatively easy to spot it on a runner level and fail hard. 
It could still be a valid to call metadata multiple times if one has a custom code with multiple metadata calls to do a setup at the start of file.

**Comment by MykolaGolubyev** [comment]
Actually :) I think putting ‘owner’ in front of each scenario may work as an user would expect. I will try it in a couple of hours.

**Comment by tsiq-karold** [comment]
>Actually :) I think putting ‘owner’ in front of each scenario may work as an user would expect. I will try it in a couple of hours.

If this is true then I feel less strongly about single call vs. multiple.  I think I still prefer multiple from a logical grouping perspective though that could be accomplished with wrappers/util functions provided by users.

**Comment by MykolaGolubyev** [comment]
I vote for single call as well. The reason my examples had a single call is due to original requirements.

**Comment by tsiq-karold** [comment]
So when I said multiple, I meant single call 😁

**Comment by MykolaGolubyev** [comment]
Ok :) single call would be my call as well (see?)

**Comment by tsiq-brian** [comment]
Upgraded webtau and was able to add in metadata. Haven't used the `afterLastTestStatement` hook yet, but I assume it'll work.

Works perfectly! Thanks @MykolaGolubyev !

I think it's safe to close this issue.

**Comment by MykolaGolubyev** [comment]
Thanks @tsiq-brian
